android/src/main/java/org/reactnative/camera/tasks/BarCodeScannerAsyncTask.java

```diff
@@ -60,81 +38,32 @@ protected Result doInBackground(Void... ignored) {
     }
 
     Result result = null;
-    /**
-     * mCameraViewWidth and mCameraViewHeight are obtained from portait orientation
-     * mWidth and mHeight are measured with landscape orientation with Home button to the right
-     * adjustedCamViewWidth is the adjusted width fromt the Aspect ratio setting 
-     */
-    int adjustedCamViewWidth = (int) (mCameraViewHeight / mRatio);
-    float adjustedScanY = (((adjustedCamViewWidth - mCameraViewWidth) / 2) + (mScanAreaY * mCameraViewWidth)) / adjustedCamViewWidth;
-    
-    int left = (int) (mScanAreaX * mWidth);
-    int top = (int) (adjustedScanY * mHeight);
-    int scanWidth = (int) (mScanAreaWidth * mWidth);
-    int scanHeight = (int) (((mScanAreaHeight * mCameraViewWidth) / adjustedCamViewWidth) * mHeight);
-
+    boolean mLimitScanArea = mScanArea.getLimitScanArea();
+    int scanWidth = mLimitScanArea ? mScanArea.getCropArea("width") : mWidth;
+    int scanHeight = mLimitScanArea ? mScanArea.getCropArea("height") : mHeight;
+    int left = mLimitScanArea ? mScanArea.getLeft() : 0;
+    int top = mLimitScanArea ? mScanArea.getTop() : 0;
+    PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
+      mImageData,
+      mHeight,
+      mWidth, 
+      mHeight - scanHeight - top,
+      left,
+      scanHeight,
+      scanWidth,
+      false
+    );
+    BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(source));
```

Comment:
@sibelius @bqwang Thanks a lot. I believe the logic is still very complex, in fact I'm reading the [PlanarYUVLuminanceSource][1] documentation and I notice now that the class is pretty simple

```java
PlanarYUVLuminanceSource(
     byte[] yuvData, // the image taken
     int dataWidth, // the width of the original picture containing barcode
     int dataHeight, // the heigh of the original picture containing barcode
     // BELOW are the cropping parameters, requires x,y starting point and width/height
     int left, // x coordinates of the top left corner starting point
     int top,  // y coordinates of the top left corner starting point
     int width, // width to crop image in this direction --->>>>
     int height, // height to crop image in this direction    V
     boolean reverseHorizontal
) 
```
we specify the left top corner, width and height for cropping. Seems to me pretty simple api and `rectOfInterest` should just pass those 4 parameters as specified below. (`x` is `left`, `y` is `top`)

```javascript
rectOfInterest={{ x: scanAreaX, y: scanAreaY, width: scanAreaWidth, height: scanAreaHeight }}
```

I believe `react-native-camera` should just expose the [zxing][2] to the `Javascript` user. The user should handle the rest. The advanteges would be:

1) We base our API on the official Android/Java apis/docs
2) We have less code
3) We have less bugs and manteinance burden. Is the responsability of the developer to handle all the remaning logic.

Took me some time to figure this out, but your comment helped me as I felt that I could not make any more improvements excepts from a totally new starting point. 

This means 90% of the previous functionality needs to be deleted, because the entire logic inside the [ScanArea](https://github.com/react-native-community/react-native-camera/blob/b258e447a8a9baf6a7c899d289c2534c53c88745/android/src/main/java/org/reactnative/camera/utils/ScanArea.java) class is just built on top of [PlanarYUVLuminanceSource][1] and superflous. 

Thanks a lot

[1]: https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html
[2]: https://github.com/zxing/zxing

Reply:
This works with my phone. `mImageData` is the original portrait image rotated to landscape (see `rotateImage` method bottom of the page).
```java
mImageData = rotateImage(imageData,mWidth, mHeight);
```
this explains why dataWidth and dataHeight are inverted (as if the image was landscape instead of portrait). The same applies for left, top, width and height cropping parameters.

```java
    PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
      mImageData, // byte[] yuvData, the image taken
      1440, // int dataWidth, the width of the original picture containing barcode
      1080, // int dataHeight, the heigh of the original picture containing barcode
       // BELOW are the cropping parameters, requires x,y starting point and width/height
       266, // int left, x coordinates of the top left corner starting point
       192, // int top,  y coordinates of the top left corner starting point
       533, // int width, width to crop image in this direction --->>>>
       666, // int height, height to crop image in this direction    V
       false, // boolean reverseHorizontal
    );
```
The `rotateImage()` method flips the image from portrait to landscape.

```java 

private byte[] rotateImage(byte[]imageData,int width, int height) {
  byte[] rotated = new byte[imageData.length];
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      rotated[x * height + height - y - 1] = imageData[x + y * width];
    }
  }
  return rotated;
}
```

Reply:
I took [a picture][1] with the Camera App and opened with Image Editing Tools (gimp).
The `this.camera.takePictureAsync()` callback returns `data` with height 4160 and width 2080 pixels. 

I [converted](https://www.freefileconvert.com/jpg-yuv) the picture to `YUV` format

I display the image on https://rawpixels.net/

1) browse and choose the yuv file
2) set width 2080 and height 4160

The image shows exactly like the original but distorted colors.

Now I am inspecting the pixel using [`getMatrix`](https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html#getMatrix--)

[1]: https://github.com/react-native-community/react-native-camera/pull/2766

Reply:
zxing library includes [a demo](https://github.com/zxing/zxing/blob/a65631d0b643e778b8b1c1975e4f18f35e401fa4/android/src/com/google/zxing/client/android/camera/CameraManager.java#L319-L325) for this functionality

Reply:
Test performed with another phone in lascape mode. The image data is **not rotated**.
```java
     PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
       mImageData,
       1520,
       1140,
       400,
       400,
       800,
       400, 
       false
    );
```

Reply:
Portrait 
```java
     PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
       rotateImage(mImageData, mWidth, mHeight),
       height, 
       width,
       top,
       left,
       height,
       width,
       false
    );
```
Landscape
```java
     PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
       mImageData,
       width,
       height, 
       left,
       top,
       width,
       height,
       false
    );
```